### PR TITLE
New interfaces for installers to run tasks on renew verb

### PR DESCRIPTION
**Pull request by joohoi**
New branch forked from #5334 only difference being the removal of `ServerTLSConfigurationUpdater`. Description of the referenced PR as well as the discussion is relevant for the other parts of this.

### CODE REVIEW

File: certbot/main.py [hunk 1]
```diff
@@ -1112,6 +1112,8 @@ def renew_cert(config, plugins, lineage):
         installer.restart()
         notify("new certificate deployed with reload of {0} server; fullchain is {1}".format(
                config.installer, lineage.fullchain), pause=False)
+        # Run deployer
+        updater.run_renewal_deployer(renewed_lineage, installer, config)
```
Comment (ohemorange) [inline certbot/main.py]: Putting this both here and in `handle_renewal_request` calls it twice during the same run. `main.renew` calls `renewal.handle_renewal_request` calls `main.renew_cert`

File: certbot/plugins/selection.py [hunk 2]
```diff
@@ -146,7 +146,7 @@ def record_chosen_plugins(config, plugins, auth, inst):
          config.authenticator, config.installer)
 
 
-def choose_configurator_plugins(config, plugins, verb):
+def choose_configurator_plugins(config, plugins, verb):  # pylint: disable=too-many-branches
```
Comment (ohemorange) [inline certbot/plugins/selection.py]: nit: don't need this here

File: certbot/plugins/selection.py [hunk 3]
```diff
@@ -192,6 +192,7 @@ def choose_configurator_plugins(config, plugins, verb):
             installer = pick_installer(config, req_inst, plugins)
         if need_auth:
             authenticator = pick_authenticator(config, req_auth, plugins)
+
```
Comment (ohemorange) [inline certbot/plugins/selection.py]: nit: extra newline

File: certbot/main.py (hunk 1 above)
Comment (joohoi) [reply certbot/main.py]: I believe the functionality is correct, as this calls the `interfaces.RenewDeployer.renew_deploy()` and `renewal.handle_renewal_request`calls `interfaces.GenericUpdater.generic_updates()` through `updater.run_renewal_updaters()`. And I don't think either of them gets called twice.

File: certbot/renewal.py [hunk 4]
```diff
@@ -426,6 +427,10 @@ def handle_renewal_request(config):
                         "cert", renewal_candidate.latest_common_version()))
                     renew_skipped.append("%s expires on %s" % (renewal_candidate.fullchain,
                                          expiry.strftime("%Y-%m-%d")))
+                # Run updater interface methods
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/renewal.py]: Unfortunately not, as the `main.renew_cert` where the `updater.run_renewal_deployer` gets called, is only run when the certificate actually gets renewed.

File: certbot/updater.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import logging
+
+from certbot import errors
+from certbot import interfaces
+
+from certbot.plugins import selection as plug_sel
+
+logger = logging.getLogger(__name__)
+
+def run_renewal_updaters(config, plugins, lineage):
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/updater.py]: None at all. I think this is a leftover from the original naming of the interfaces. It's now changed to `run_generic_updaters`.

File: certbot/renewal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ohemorange) [reply certbot/renewal.py]: Hold up, I thought that the renew deployer is what's called when certs are actually renewed, no?

[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/renewal.py]: Yes, exactly. This, (`updater.run_generic_updaters`) however gets run regardless of the cert getting actually renewed.

[the remark this answers is not in the document]
Comment (ohemorange) [reply certbot/renewal.py]: Oh, I see the confusion -- I was suggested putting it in `main.renew` not `main.renew_cert`.

[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/renewal.py]: I would prefer not to, as we'd have to duplicate most of the functionality in `renewal.handle_renewal_request` in order to do so, namely fetching the renewal config files, getting lineage for them, picking and initiating the installer and finally running the updaters.

[the remark this answers is not in the document]
Comment (ohemorange) [reply certbot/renewal.py]: Ah hm I see. Ok, cool. Makes me sad but that's on the existing codebase, not this change.

File: certbot/updater.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    :returns: `None`
+    :rtype: None
+    """
+    try:
+        # installers are used in auth mode to determine domain names
+        installer, _ = plug_sel.choose_configurator_plugins(config, plugins, "certonly")
+    except errors.PluginSelectionError as e:
+        logger.info("Could not choose appropriate plugin: %s", e)
+        raise
```
Comment (ohemorange) [inline certbot/updater.py]: Hm. Maybe we don't want to `raise` here -- if a single lineage is broken, we should still be trying to update the others, no?

Comment (joohoi) [reply certbot/updater.py]: This is handled by the pre-existing try-except block in `renewal.handle_renewal_request` which calls this method: https://github.com/certbot/certbot/blob/plugin_renew_updates/certbot/renewal.py#L434-L440

However you are right that we probably shouldn't raise this anyway as the error message displayed for the user by the block above might confusing and is definitely inaccurate. It tells that there was an error with the renewal, which isn't the case in this situation.

I removed the `raise` and instead changed the logger level to `warning` and made the log message a bit more informative.

**Pull request merged**